### ISSUE TITLE
CompileNodeToValue does not evaluate the value of constant expressions

### ISSUE THREAD

**Issue by asgrim** [Roave/BetterReflection#19]
If you have a parameter with a default value of a constant (e.g. `self::WHATEVER`), we currently do not evaluate the *value* of `self::WHATEVER`, we just return `self::WHATEVER` as a string. This is incorrect, we need to (wherever possible) determine the actual value.

**Comment by asgrim** [comment]
Failing test case:

```php
public function testClassConstantUsedAsDefaultValue()
{
    $php = '<?php
        class MyClass {
            const FOO = 123;

            public function bar($baz = self::FOO);
        }
    ';

    $tree = (new Parser(new Lexer))->parse($php);

    $actualValue = (new CompileNodeToValue())
        ->__invoke($tree[0]->stmts[1]->params[0]->default);
    $this->assertSame(123, $actualValue);
}
```

**Comment by Ocramius** [comment]
Few notes:

 - constants in a class can refer to other constants in a class (recursive evaluation)
 - constants may be undefined (php does trigger a notice at runtime in those cases: http://3v4l.org/rDiY1)

**Comment by asgrim** [comment]
Also note that you could have something like:

```php
<?php
define("LOL", $argv[0]);

function foo($foo = LOL) {
 var_dump($foo);
}
```

which you can never determine statically, because $argv[0] is out of context. Note that `define` is a `Node\Expr\FuncCall`, and the second parameter is a node something like:

```
Node\Expr\ArrayDimFetch
  ->var = Node\Expr\Variable
    ->name = argv
  ->dim = Node\Scalar\LNumber
    ->value = 0
```

However, pretty much exactly the same but we **can** work this one out:

```php
<?php
define("LOL", 123);

function foo($foo = LOL) {
 var_dump($foo);
}
```

As you'd just find all the constants defined as "LOL" and see it is a scalar expression. It gets a little more complicated when you have:

```php
<?php
$foo = [1, 2, 3];
define("LOL", $foo[0]);

function foo($foo = LOL) {
 var_dump($foo);
}
```

With this you have almost no more information in the specific node, so you MUST look around in the context (the code...) and try to evaluate the expression.

```
Node\Expr\ArrayDimFetch
  ->var = Node\Expr\Variable
    ->name = foo
  ->dim = Node\Scalar\LNumber
    ->value = 0
```

Looking at the WHOLE tree however, we can see that `$foo[0]` can eventually be compiled into a scalar value, which means we can eventually define it. Basically, we need to absolutely be able to recurse through nodes, easily locate/define variable nodes in the AST to look up values etc.

Naturally, the ultimate goal is that we end up with a `Node\Scalar` instance. That's the rule. If we don't end up with a `Node\Scalar`, I think we can't evaluate the expression statically. Once we have determined how to actually search the AST for values and are able to evaluate them, we should be able to determine that we *can't* compile this statically:

```php
<?php
define('LOL', $argv[0]);

class Foo {
  const FOO_ARG = LOL;

  public function myFunc($default = self::FOO_ARG)
  {
    echo $default;
  }
}

$foo = new Foo();
$foo->myFunc();
```

but we *can* compile this statically:

```php
<?php
define('LOL', 123);

class Foo {
  const FOO_ARG = LOL;

  public function myFunc($default = self::FOO_ARG)
  {
    echo $default;
  }
}

$foo = new Foo();
$foo->myFunc();
```

This is just subtle tweak but it means it works because the node searching eventually stops at a `Node\Scalar\LNumber`. Steps to achieve:

* Given initially that we know `self::FOO_ARG` is a `Node\Expr\ClassConstFetch` (**go and fetch the class constant FOO_ARG**)
* Traverse **up** the AST to the class-level, loop through the `stmts` and locate all `Node\Stmt\ClassConst` nodes, examine `ClassConst->consts[0]->name` to find it is a `Node\Const_` called `FOO_ARG`. See that the `->value` is an instance of `Node\Stmt\ConstFetch` (**go and fetch the constant LOL**)
* Search all the AST, find all the `Node\Stmt\FuncCall` where `FuncCall->name->parts[0] === 'define'`. If the `FuncCall->args[0]->value->value === 'LOL'` then we have the correct Const definition... so examine `FunctCall->args[1]->value` and see it is a `Node\Scalar\LNumber` which we can pick the value of (`FunctCall->args[1]->value->value`)
* Note that if we cannot find the appropriate `define()` function call statement, then we abandon - we cannot evaluate the value with the given context
* Now we know that `self::FOO_ARG` evaluates to `123`

**Comment by asgrim** [comment]
```php
<?php
class Foo {
  const MY_CONST = 123;

  public function myFunc($default = self::MY_CONST)
  {
    echo $default;
  }
}
```

After a bit of discussing on IRC, probably not a good idea to get too far down this rabbit hole. Lets solve a common use case - the `Node\Expr\ClassConstFetch` as default parameter. If we encounter this node, we require a `ReflectionClass` to be given as context. Call `ReflectionClass->getConstant('MY_CONST')` (to be implemented in #7) which in turn at call-time (to be implemented in #34) will try to evaluate the node. Assuming it's a scalar or array (in this example, it's `Node\Scalar\LNumber`, so valid), then we will evaluate and return. If anything else, then just throw a `EvaluationNotImplemented` exception with a helpful message.

**Comment by asgrim** [comment]
Closed in #95
